Remove locks around ExecutionStateSampler

steveniemitz opened the pull request:
This fixes the ExecutionStateSampler portion of #22161.

There's a couple options I've tried to remove the lock there:

- Switch activeTrackers back to a concurrent set, the problem here was that deactivating a tracker could still race with the sampler thread and double-sample a sampler. I'm not sure if this is a big enough deal to worry about?
- Rather than worry about synchronization at all, push all add/remove/sample operations into a queue and have a single thread consuming from it and maintaining activeTrackers and doing the sampling. Doing so removes the race above and simplifies the logic a lot (I think). I used the LMAX Disruptor queue to handle the multi-consumer-single-producer setup here, but you could probably use a built-in java concurrent blocking queue for this as well.

Benchmark results are as follows:

![image](https://user-images.githubusercontent.com/1882981/178150077-64f9aac3-45bd-428f-9240-74a5951540c2.png)

The machine I was running these tests on has 24 logical cores, which is why we see a big drop in throughput between 20 -> 50 threads.

Both the LMAX and stock (synchronized) implementations have a significant throughput dropoff after the number of cores is saturated.  The synchronized one due to lock contention, and the LMAX one due to the disruptor queue filling up (because the worker thread gets scheduled out).

The ConcurrentSet implementation seems the best here, with no scalability dropoff after core saturation.

One thing to note: I was still unable to reproduce the significant blocking I saw in a real dataflow job with the JMH benchmarks, even at 100 threads the contention was very low (~1 second of blocked threads over a 1 minute sample, vs 30 seconds over a 1 minute sample in dataflow).

R: @lukecwik 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

lukecwik commented:
Why don't we get https://github.com/apache/beam/pull/22103 in so you can re-use the benchmarks there which would show up meaningful differences in the LMAX, concurrent set, and active tracker solutions.

If you don't got the time to review https://github.com/apache/beam/pull/22103, I can try to find another person.

steveniemitz commented:
> Why don't we get #22103 in so you can re-use the benchmarks there which would show up meaningful differences in the LMAX, concurrent set, and active tracker solutions.
> 
> If you don't got the time to review #22103, I can try to find another person.

sounds good to me, I can look at it right now

steveniemitz commented:
I updated this to use a ConcurrentHashMap-backed-set, which seems to have the best performance of the 3 options (I updated the description as well).  The JMH benchmarks were super helpful here.

lukecwik commented on runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/ExecutionStateTracker.java:
```diff
@@ -298,7 +304,17 @@ public long getNextLullReportMs() {
 
-  protected void takeSample(long millisSinceLastSample) {
+  void takeSample(long millisSinceLastSample) {
+    if (SAMPLING_UPDATER.compareAndSet(this, 0, 1)) {
```
Do we need sampling/SAMPLING_UPDATER because you wanted to be able to have multiple threads activating/deactivating states within a single tracker at the same time?

Note there is only one ExecutionStateSampler thread and each tracker is associated with a single bundle at a time. To emulate this behavior we would want to have one sampler that is shared by N trackers each with their own states where N is the number of concurrent threads you want to use.

steveniemitz replied:
There's a race now between `removeTracker` removing the tracker from the concurrent set (and them sampling it) and the sampling thread enumerating it and sampling it.  Previously the synchonrized block would prevent that race.  This is essentially pushing the lock down to the individual tracker instances to prevent it being sampled concurrently.

lukecwik replied:
Are you saying that there is another thread that does the sampling that isn't the ExecutionStateSampler thread?

steveniemitz replied:
Correct, it's sampled in `removeTracker` as well https://github.com/apache/beam/blob/534e99eae37e34760baa7beb432a18dc4cd148a5/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/ExecutionStateSampler.java#L166

lukecwik replied:
Thanks, I missed that.

lukecwik commented:
> I updated this to use a ConcurrentHashMap-backed-set, which seems to have the best performance of the 3 options (I updated the description as well). The JMH benchmarks were super helpful here.

Finally it would be good to get some numbers for what the updated benchmark gets before and after your change.

steveniemitz commented:
> Finally it would be good to get some numbers for what the updated benchmark gets before and after your change.

It's in the description, or were you thinking of some other data?

steveniemitz commented on sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/control/ExecutionStateSamplerBenchmark.java:
```diff
@@ -103,33 +104,38 @@ public void tearDown() {
   @Benchmark
-  @Threads(1)
-  public void testTinyBundleRunnersCoreStateSampler(RunnersCoreStateSampler state)
+  @Threads(10)
```
oh I didn't mean to actually commit this, I'll revert

lukecwik replied:
I think adding the threads seems worthwhile to show the cost of contention.

lukecwik replied:
Did we mean to have 512 threads for the tiny bundles and 16 for the large bundles or are you still meaning to revert this?

steveniemitz replied:
oh I was going to leave it at 10 for everything, I can switch it up though if you want.

lukecwik commented on sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/control/ExecutionStateSamplerBenchmark.java:
```diff
@@ -103,33 +104,38 @@ public void tearDown() {
-  public void testTinyBundleRunnersCoreStateSampler(RunnersCoreStateSampler state)
+  @Threads(10)
+  @Fork(1)
+  public void testTinyBundleRunnersCoreStateSampler(RunnersCoreStateSampler state, Blackhole bh)
```
Did you see that we needed the blackhole?

steveniemitz replied:
I figured it'd be better to be safe than sorry, it also adds a little non-trivial work to the loop which is nice for the benchmark.

lukecwik replied:
sgtm

steveniemitz commented:
Run Java PreCommit

steveniemitz commented:
Run Java PreCommit

steveniemitz commented:
The precommit is failing with "cannot find symbol" warnings in unrelated code, it looks like its failed on other PRs for the same reason as well, so seems unrelated to this change.

lukecwik commented:
Run Java PreCommit

lukecwik commented on sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/control/ExecutionStateSamplerBenchmark.java:
```diff
@@ -138,33 +142,37 @@ public void testTinyBundleHarnessStateSampler(HarnessStateSampler state) throws
+  @Threads(10)
+  public void testLargeBundleHarnessStateSampler(HarnessStateSampler state, Blackhole bh)
+      throws Exception {
     state.tracker.start("processBundleId");
```
We don't want to have this tracker be shared across threads.

lukecwik commented on sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/control/ExecutionStateSamplerBenchmark.java:
```diff
@@ -103,33 +103,37 @@ public void tearDown() {
+  public void testTinyBundleRunnersCoreStateSampler(RunnersCoreStateSampler state, Blackhole bh)
       throws Exception {
-    state.tracker.activate();
+    ExecutionStateTracker tracker = new ExecutionStateTracker(state.sampler);
```
We should create a JMH `@State` called `RunnersCoreStateTracker` at thread scope. Then in the setup method take in a parameter of the RunnersCoreStateSampler.

You can see an example of what I mean here: https://github.com/openjdk/jmh/blob/1f2befef92c3eb1466124f37d00f496b4105d3c5/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_29_StatesDAG.java#L134

This will allow us to move the initialization outside of the benchmark into state. 

Ditto on creating `HarnessStateTracker` as well.

lukecwik reviewed (changes requested):
I think your code update looks good, only concern is around completing the changes to the benchmark.

lukecwik commented on sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/control/ExecutionStateSamplerBenchmark.java:
```diff
@@ -41,6 +41,16 @@
+
   @State(Scope.Benchmark)
   public static class RunnersCoreStateSampler {
     public final ExecutionStateSampler sampler = ExecutionStateSampler.newForTest();
```
I think these states need to be moved to RunnersCoreStateTracker like you did for HarnessStateTracker.

We would only ever expect to have one thread activating/deactivating the state. The race condition that you pointed out was between the sampling thread and one (not many) processing thread.

steveniemitz replied:
ah yeah, makes sense!

steveniemitz commented:
any idea how to re-run these failed actions?

lukecwik commented:
I restarted them.

You click `Details` on the far right and in the top left you can click on the `Rerun jobs` button and select `Rerun failed jobs`.

steveniemitz commented:
> You click `Details` on the far right and in the top left you can click on the `Rerun jobs` button and select `Rerun failed jobs`.

ah ok I must not have permissions to do that.

lukecwik commented:
> > You click `Details` on the far right and in the top left you can click on the `Rerun jobs` button and select `Rerun failed jobs`.
> 
> ah ok I must not have permissions to do that.

It only shows up when there are failed jobs for me. You might not see it right now because I restarted them or you could be right and its a permissions thing.
